### PR TITLE
[Scheduler] Throw an exception when no dispatcher has been passed to a Schedule

### LINKED ISSUE #53722: [Scheduler] Using before, after or onFailure methods causes unhandled exception when not configured
### Symfony version(s) affected

6.4

### Description

According to the documentation at: 
https://symfony.com/doc/current/scheduler.html#a-strategic-event-handling
image

Using those functions like that does not work and results in a 500 error
```
{"message":"Uncaught Error: Call to a member function addListener() on null","context":{"exception":{"class":"Error","message":"Call to a member function addListener() on null","code":0,"file":"/app/vendor/symfony/scheduler/Schedule.php:144"}},"channel":"php","extra":{},"severity":"CRITICAL","time":"2024-02-01T13:22:04.559+01:00"}
```

### How to reproduce

```
     // ...
    public function getSchedule(): Schedule
    {
        $this->schedule ??= (new Schedule())
            ->with(...$this->getGeneralMessages(), ...$this->getProductionMessages())
            ->before(function (PreRunEvent $event) {
                $this->logger?->info('Before Event');
            })
            ->after(function (PostRunEvent $event) {
                $this->logger?->info('After Event');
            })
            ->onFailure(function (FailureEvent $event) {
                $this->logger?->info('Failure Event');
                $this->logger?->info($event->getError()->getMessage());
            })
            ->lock($this->lockFactory->createLock('cron-cache'))
            ->stateful($this->cache);

        return $this->schedule;
    }
```

### Possible Solution

Either update documentaiton, add error handling and throw our own exception or fail silently and log an error?

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Issues        | Fix #53722 
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/Scheduler/Schedule.php [hunk 1]
```diff
@@ -141,20 +141,32 @@ public function getSchedule(): static
...
+
         $this->dispatcher->addListener(PreRunEvent::class, $listener, $priority);
 
         return $this;
     }
 
     public function after(callable $listener, int $priority = 0): static
     {
+        if (!$this->dispatcher) {
+            throw new LogicException('To register a "before" listener, you need to set an event dispatcher on the Schedule.');
```
Comment (xabbuh) [inline src/Symfony/Component/Scheduler/Schedule.php]:
```suggestion
            throw new LogicException('To register an "after" listener, you need to set an event dispatcher on the Schedule.');
```

File: src/Symfony/Component/Scheduler/Schedule.php [hunk 2]
```diff
@@ -141,20 +141,32 @@ public function getSchedule(): static
...
+
         $this->dispatcher->addListener(PostRunEvent::class, $listener, $priority);
 
         return $this;
     }
 
     public function onFailure(callable $listener, int $priority = 0): static
     {
+        if (!$this->dispatcher) {
+            throw new LogicException('To register a "before" listener, you need to set an event dispatcher on the Schedule.');
```
Comment (xabbuh) [inline src/Symfony/Component/Scheduler/Schedule.php]:
```suggestion
            throw new LogicException('To register a "failure" listener, you need to set an event dispatcher on the Schedule.');
```

File: src/Symfony/Component/Scheduler/Schedule.php (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Component/Scheduler/Schedule.php]: Fixed

File: src/Symfony/Component/Scheduler/Schedule.php (hunk 2 above)
Comment (fabpot) [reply src/Symfony/Component/Scheduler/Schedule.php]: Fixed

File: src/Symfony/Component/Scheduler/Schedule.php [hunk 3]
```diff
@@ -141,20 +141,32 @@ public function getSchedule(): static
 
     public function before(callable $listener, int $priority = 0): static
     {
+        if (!$this->dispatcher) {
+            throw new LogicException(sprintf('To register a "%s" listener, you need to set an event dispatcher on the Schedule.', __METHOD__));
```
Comment (xabbuh) [inline src/Symfony/Component/Scheduler/Schedule.php]: Maybe reword this a bit?

```suggestion
            throw new LogicException(sprintf('To register a listener with "%s()" you need to set an event dispatcher on the Schedule.', __METHOD__));
```

Comment (fabpot) [reply src/Symfony/Component/Scheduler/Schedule.php]: LGTM

**Review by xabbuh — APPROVED** [review]
